File: synthetic_data_generator/generator_code/toolboxTABLESgenerator.R

### toolbox for TABLE GENERATION

# --------------------------------------------------------------------------------------------------
# STEP 1: Get Names and Values
# Description: Retrieves the 'names' and 'values' for a given attribute code from semantic data.
# get the names and values of attribute code from semantic data
getNamesValues <- function(semanticData,code){
  for (i in 1: length(semanticData)){
    if (semanticData[[i]]$code == code) { 
      res <- NULL
      res$names = semanticData[[i]]$names
      res$values = semanticData[[i]]$values
      return(res)
    }
  }
  return(NULL)
}

#--Test Data------------------------------------------------------------------------------------------
# semanticData <- list(
#   list(code = "x1", names = c("country"), values = c("USA", "UK")),
#   list(code = "x2", names = c("year"), values = c(2020, 2021))
# )
# getNamesValues(semanticData, "x1")


# generate samples for the value column
# ensure integer values are distinct
# valueCode can be a code for values as in semantics or directly an interval of values
# use dot like in 2.3 and 24.0 to identify a real number and no dot like 24 to identify integers
# any interval containing at least a real as min or max will generate real numbers uniformly sampled in that interval

# --------------------------------------------------------------------------------------------------
# STEP 2: Make Distinct
# Description: Ensures uniqueness by adding a small offset to duplicate values.

# generate distinct values from a list of values by adding epsilon offsetwith numDigits decimal values
# to each duplicate, then randomize their order
makeDistinct <- function(vals,numDigits){
  #print("makeDistinct")
  #print(vals)
  while (length(unique(vals))!=length(vals)){
    # there are still duplicates
    svals = sort(vals) 
    for (i in which(diff(svals)==0)){
      # add the smallest possible value to each duplicate
      svals[i+1] = svals[i+1]+10^(-numDigits)
    }
    vals = svals
  }
  
  # return randomize order
  return(sample(vals)) 
}


# --------------------------------------------------------------------------------------------------
# STEP 3: Get Sample Values (Numeric)
# Description: Generates samples from semantic ranges (integer or real values).
getSampleValues <- function(semanticValues,valueCode,numSample,numDigits=2){
  
  if (length(valueCode) == 1){ # valueCode is a name of a code like "int100" in semanticValues to generate values 
    for (cn in names(semanticValues)){
      if (cn == valueCode) { 
        rangeVal = unlist(semanticValues[[cn]])
        sv1 = semanticValues[[cn]][[1]]
        sv2 = semanticValues[[cn]][[2]]
        
        if (is.integer(sv1) && is.integer(sv2)) {
          minv = rangeVal[1]
          maxv = max(minv+numSample,rangeVal[2])
          vals = sample(minv:maxv,numSample,replace = FALSE) # ensure numSample distinct values are generated
        } else {
          vals = round(runif(numSample,min = rangeVal[1],max = rangeVal[2]), digits = numDigits)
        }
      }
    }
  }else{
    
    # valueCode is directly a list of two values min and max: c(m,M) given in the field "values" of the table json format
    rangeVal = unlist(valueCode)
    sv1 = valueCode[[1]]
    sv2 = valueCode[[2]]
    
    if (is.integer(sv1) && is.integer(sv2)) {
      minv = rangeVal[1]
      maxv = max(minv+numSample,rangeVal[2])
      vals = sample(minv:maxv,numSample,replace = FALSE) # ensure numSample distinct values are generated
    } else {
      vals = round(runif(numSample,min = rangeVal[1],max = rangeVal[2]), digits = numDigits)
    }
  }
  vals = makeDistinct(vals,numDigits)
  return(vals)
}


#--Test Data------------------------------------------------------------------------------------------
# semanticValues <- list(int5 = c(1L, 5L))
# set.seed(123)
# getSampleValues(semanticValues, "int5", numSample = 3, numDigits = 0)


# --------------------------------------------------------------------------------------------------
# STEP 4: Get Values From Names
# Description: Extracts attribute values based on inclusion/exclusion filters.
# get the values of attribute from their names and filters
getValuesFromNames <- function(namesAndValues,valKeep,valRemove){
  speStr = "!@!" # special sep to distinguish numbers of the names and numbers added when unlisting
  nam = namesAndValues$names
  val = namesAndValues$values
  if (length(nam)>1){
    
    # if names contain numbers, add a character at the end to allow proper clean up
    for (ii in 1:length(val)){
      curname = names(val[[ii]])
      names(val[[ii]]) = paste0(curname,speStr)
    }
    
    # TREE 
    uval=unlist(val) # it adds numbers at the end of the names to avoid duplicates
    # remove numbers after speStr
    cleanuval = gsub(speStr, '',gsub(paste0(speStr,'[[:digit:]]+'), '',names(uval)))
    cleanvalues = paste0(cleanuval,'.',uval)
    sepvalues = strsplit(cleanvalues,split = "\\.")
    
    Ltmp <- NULL
    # apply filter
    noFilter = TRUE
    # keep (priority over remove list, remove list will not be considered if keep list is used)
    if (length(valKeep)>0) {
      for (fi in 1:length(valKeep)){
        keepval = valKeep[fi]
        for (gi in 1:length(sepvalues)){
          if (any(sepvalues[[gi]] == keepval)) Ltmp <- rbind(Ltmp, paste(sepvalues[[gi]],collapse="."))
        }
      }
      noFilter = FALSE
    }
    # remove (remove list ONLY considered if keep list not present
    if (length(valRemove)>0 && length(valKeep)==0) {
      for (fi in 1:length(valRemove)){
        removeval = valRemove[fi]
        for (gi in 1:length(sepvalues)){
          if (all(sepvalues[[gi]] != removeval)) Ltmp <- rbind(Ltmp, paste(sepvalues[[gi]],collapse="."))
        }
      }
      noFilter = FALSE
    }
    
    
    # generate all data
    if (noFilter) {
      Ltmp <- cleanvalues
    }
    
    
  } else {
    # NOT A TREE, SINGLE ATTRIBUTE
    uval=unlist(val)
    
    Ltmp <- NULL
    # apply filter
    noFilter = TRUE
    # keep (priority over remove list, remove list will not be considered if keep list is used)
    if (length(valKeep)>0) {
      for (fi in 1:length(valKeep)){
        keepval = valKeep[fi]
        for (gi in 1:length(uval)){
          if (uval[gi] == keepval) Ltmp <- rbind(Ltmp, uval[gi])
        }
      }
      noFilter = FALSE
    }
    # remove (remove list ONLY considered if keep list not present
    if (length(valRemove)>0 && length(valKeep)==0) {
      for (fi in 1:length(valRemove)){
        removeval = valRemove[fi]
        for (gi in 1:length(uval)){
          if (uval[gi] != removeval) Ltmp <- rbind(Ltmp, uval[gi])
        }
      }
      noFilter = FALSE
    }
    
    
    # generate all data
    if (noFilter) {
      Ltmp <- uval
    }
    
  }
  
  return(Ltmp)
}


# -- Test Case --
# namesAndValues <- list(
#   names = c("country", "year"),
#   values = list(
#     country = c("USA", "UK"),
#     year = c(2020, 2021)
#   )
# )
# print(getValuesFromNames(namesAndValues, c(), c()))
# print(getValuesFromNames(namesAndValues, c("UK"), c()))
# print(getValuesFromNames(namesAndValues, c(), c("USA")))


# --------------------------------------------------------------------------------------------------
# STEP 5: Sample Values from List
# Description: Randomly selects consecutive values from a list.

# sample values from filtered list
# use valSample = [0] to pick all existing values without sampling
# pick between valSample[1] and valSample[2] consecutive values in valList list of values
# for instance: in [1,2,3,4,5] pick [2,4] can generate at random either: [1,2],[2,3],[3,4],[4,5],[1,2,3],[2,3,4],[3,4,5],[1,2,3,4],[2,3,4,5]
sampleValues <- function(valList,valSample){
  # sample
  Ltmp = valList
  if (length(valSample)==2) {
    ## DO SAMPLING [n,m] means: pick between n and m consecutive values in current list of values
    n = valSample[1]
    m = valSample[2]
    numVal = length(valList)
    n = max(1,min(n,numVal)) # make sure 1 <= n <= numVal
    m = max(n,min(m,numVal)) # make sure n <= m <= numVal
    if (n==m) { siz = n } else { siz = sample(n:m,1) } # rand number between n and m
    pos = sample(1:(numVal-siz+1),1) # rand position between 1 and numVal-siz+1
    ind = pos:(pos+siz-1) # list of index to sample from current list of values
    Ltmp <- valList[ind]
  }
  
  isAggPossible = "true"
  if (length(Ltmp)==1) isAggPossible = "false"
  
  res <-NULL
  res$isAggPossible <- isAggPossible
  res$listAttrNames <- Ltmp
  
  return(res)
}
# -- Test the function --
# set.seed(123)
# valList <- c("A", "B", "C", "D", "E")
# valSample <- c(2, 3)

# print(sampleValues(valList, valSample))



# --------------------------------------------------------------------------------------------------
# STEP 6: Sample Values From Hierarchical Names
# Description: Samples attribute names across multiple hierarchy levels.

## generate samples follwoing valSample (use sampleValues internally)
## with listCompositeAttrNames a list of attribute names made of nested levels separated by "."
## sampling is done on each branch at each level independently
sampleValuesFromHierarchy <- function(listCompositeAttrNames,valSample){
  
  
  # it has added '.' between level names
  LattrAllLevels = strsplit(listCompositeAttrNames,split="\\.")
  
  depthAttr = length(LattrAllLevels[[1]])
  
  # for parent level, sample which to keep
  MattrAllLevelsINI = matrix("",length(LattrAllLevels),depthAttr)
  
  for (ii in 1:depthAttr){
    # get parent
    for (jj in 1:length(LattrAllLevels)){
      MattrAllLevelsINI[jj,ii] = LattrAllLevels[[jj]][ii] 
    }
  }
  MattrAllLevels = MattrAllLevelsINI
  # sample level 1
  attLev = MattrAllLevels[,1]
  uAttLev = unique(attLev)
  
  
  if (length(valSample)==2) {
    
    res = sampleValues(uAttLev,valSample)
    splAttLev = res$listAttrNames
    
    # record if it is possible to aggregate several attribute values at that level
    isAggPossible <- NULL
    isAggPossible[1] = res$isAggPossible
    #if (length(splAttLev)==1) isAggPossible[1] = "false" # a single attribute value at that level, no aggregate
    
    MattrAllLevels = MattrAllLevels[attLev %in% splAttLev,]
    
    for (ii in 2:depthAttr){
      upar = unique(MattrAllLevels[,1])
      lupar = length(upar)
      subMattrAllLevels = NULL
      isAggPossible[ii] = "true"
      for (ipar in 1:lupar){
        curpar = upar[ipar]
        attLev = MattrAllLevels[MattrAllLevels[,1]==curpar,,drop = F]
        uAttLev = unique(attLev[,2])
        res = sampleValues(uAttLev,valSample)
        splAttLev = res$listAttrNames
        if (res$isAggPossible == "false") isAggPossible[ii] = "false"
        subMattrAllLevels = rbind(subMattrAllLevels, attLev[attLev[,2] %in% splAttLev,])
      }
      # concatenate current and parent column
      Lcollapsed = NULL
      for (ipr in 1:nrow(subMattrAllLevels)){
        Lcollapsed[ipr] = paste0(subMattrAllLevels[ipr,1:2],collapse=".")
      }
      if (ncol(subMattrAllLevels)>2){
        MattrAllLevels=cbind(Lcollapsed,subMattrAllLevels[,3:ncol(subMattrAllLevels)])
      } else {
        MattrAllLevels=matrix(Lcollapsed,nrow(subMattrAllLevels),1)
        break
      }
    }
    res <-NULL
    res$isAggPossible <- paste0(isAggPossible,collapse=".")
    res$listAttrNames <- c(MattrAllLevels)
    return(res)
    
  } else {
    
    # record if it is possible to aggregate several attribute values at that level
    isAggPossible <- NULL
    isAggPossible[1] = "true"
    if (length(uAttLev)==1) isAggPossible[1] = "false" # a single attribute value at that level, no aggregate
    
    
    for (ii in 2:depthAttr){
      upar = unique(MattrAllLevels[,1])
      lupar = length(upar)
      subMattrAllLevels = NULL
      isAggPossible[ii] = "true"
      for (ipar in 1:lupar){
        curpar = upar[ipar]
        attLev = MattrAllLevels[MattrAllLevels[,1]==curpar,,drop = F]
        uAttLev = unique(attLev[,2])
        if (length(uAttLev)==1) isAggPossible[ii] = "false"
        subMattrAllLevels = rbind(subMattrAllLevels, attLev[attLev[,2] %in% uAttLev,])
      }
      # concatenate current and parent column
      Lcollapsed = NULL
      for (ipr in 1:nrow(subMattrAllLevels)){
        Lcollapsed[ipr] = paste0(subMattrAllLevels[ipr,1:2],collapse=".")
      }
      if (ncol(subMattrAllLevels)>2){
        MattrAllLevels=cbind(Lcollapsed,subMattrAllLevels[,3:ncol(subMattrAllLevels)])
      } else {
        MattrAllLevels=matrix(Lcollapsed,nrow(subMattrAllLevels),1)
        break
      }
    }
    res <-NULL
    res$isAggPossible <- paste0(isAggPossible,collapse=".")
    res$listAttrNames <- listCompositeAttrNames
    return(res)
  }
}

# -- Test Case --
# compositeAttrs <- c("Asia.India.Delhi", "Asia.India.Mumbai", "Europe.Germany.Berlin")
# valSample <- c(1, 2)
# set.seed(123)
# print(sampleValuesFromHierarchy(compositeAttrs, valSample))


# --------------------------------------------------------------------------------------------------
# STEP 7: Get Attribute Index from Code
# Description: Returns the index of the attribute code in the attribute list.

# get index of col/row attribute having a certain code (return the first one if a list of codes is given) 
getAttrFromCodes <- function(attrList,codes){ 
  for (i in 1:length(attrList)) if (is.element(attrList[[i]]$code,codes)) return(i)
} 


# -- Test Case --
# attrList <- list(
#   list(code = "x1", names = c("Country")),
#   list(code = "x2", names = c("Year")),
#   list(code = "x3", names = c("City"))
# )
# print(getAttrFromCodes(attrList, "x2"))   # Expected: 2
# print(getAttrFromCodes(attrList, "x99"))  # Expected: NULL


# --------------------------------------------------------------------------------------------------
# STEP 8: Get Codes from Name
# Description: Maps attribute names to their codes based on semantic metadata.

getCodesFromName <- function(allSemanticAttributes,attrName){
  attrData <- allSemanticAttributes$data
  codes <- NULL
  for (i in 1:length(attrData)) if (is.element(attrName,unlist(attrData[[i]]$names))) codes <- c(codes,attrData[[i]]$code)
  return(codes)
}

# -- Test Case --
# semanticData <- list(
#   data = list(
#     list(code = "x1", names = c("Country")),
#     list(code = "x2", names = c("Year")),
#     list(code = "x3", names = c("City"))
#   )
# )
# print(getCodesFromName(semanticData, "Year"))     # Expected: "x2"
# print(getCodesFromName(semanticData, "Unknown"))  # Expected: NULL

#  Final Test (R) — Full Flow ----------------------------------
# Load your toolbox
source("toolboxTablesgenerator.R")

# Step 1: Mock semanticData
semanticData <- list(
  list(code = "x1", names = c("Country"), values = list(Country = c("USA", "UK", "Qatar"))),
  list(code = "x2", names = c("Year"), values = list(Year = c(2020, 2021, 2022)))
)

# Step 1: Get names/values
nv <- getNamesValues(semanticData, "x1")

# Step 2 & 3: Filter values
filtered_vals <- getValuesFromNames(nv, valKeep = c(), valRemove = c("UK"))

# Step 4: Sample flat values
sample_cfg <- c(1, 2)
sampled <- sampleValues(filtered_vals, sample_cfg)

# Step 5: Hierarchical test
compositeAttrs <- c(
  "Asia.Qatar.Doha", "Asia.Qatar.AlRayyan", "Asia.UAE.Dubai",
  "Europe.UK.London", "Europe.UK.Manchester"
)
set.seed(123)
hct_result <- sampleValuesFromHierarchy(compositeAttrs, sample_cfg)

# Step 6: Get index of attribute
attr_index <- getAttrFromCodes(semanticData, "x2")

# Step 7: Get code from name
semanticDataObj <- list(data = semanticData)
code_result <- getCodesFromName(semanticDataObj, "Country")

# Show results
cat("🔹 Filtered:", filtered_vals, "\n")
print(sampled)
print(hct_result)
cat("🔹 Attribute Index for 'x2':", attr_index, "\n")
cat("🔹 Code for 'Country':", code_result, "\n")

File: synthetic_data_generator/python_generator_code/toolbox_tables_generator.py
import numpy as np
import random
#DONE
# -----------------------------------------------
# STEP 1: Extract names and values by code
# -----------------------------------------------
# Returns the 'names' and 'values' for a given semantic attribute code.

def get_names_values(semantic_data, code):
    for entry in semantic_data:
        if entry.get('code') == code:
            return {
                'names': entry.get('names', []),
                'values': entry.get('values', [])
            }
    return {'names': [], 'values': []}

# # Test STEP 1
# semantic_data = [
#     {"code": "x1", "names": ["country"], "values": ["USA", "UK"]},
#     {"code": "x2", "names": ["year"], "values": [2020, 2021]}
# ]
# print(get_names_values(semantic_data, "x1"))
# # → {'names': ['country'], 'values': ['USA', 'UK']}


# -----------------------------------------------
# STEP 2: Sample values from numeric range
# -----------------------------------------------
# Uses named or direct range codes and ensures uniqueness using make_distinct().

def make_distinct(vals, num_digits=2):
    epsilon = 10 ** -num_digits
    vals = list(vals)
    while len(set(vals)) != len(vals):
        vals.sort()
        for i in range(len(vals) - 1):
            if vals[i] == vals[i+1]:
                vals[i+1] += epsilon
    np.random.shuffle(vals)
    return vals

def get_sample_values_numeric(semantic_values: dict, value_code, num_sample: int, num_digits: int = 2) -> list:
    if isinstance(value_code, str) and value_code in semantic_values:
        lo, hi = semantic_values[value_code]
    elif isinstance(value_code, (list, tuple)) and len(value_code) == 2:
        lo, hi = value_code
    else:
        raise ValueError(f"Invalid value_code: {value_code}")

    if float(lo).is_integer() and float(hi).is_integer():
        lo_i, hi_i = int(lo), int(hi)
        minv = lo_i
        maxv = max(minv + num_sample, hi_i)
        pool = list(range(minv, maxv + 1))
        vals = np.random.choice(pool, size=num_sample, replace=False).tolist()
    else:
        vals = np.round(np.random.uniform(lo, hi, size=num_sample), num_digits).tolist()

    return make_distinct(vals, num_digits)

# # Test STEP 2
# semantic_values = {"int5": [1, 5]}
# np.random.seed(123)
# print(get_sample_values_numeric(semantic_values, "int5", 3, num_digits=0))
# # → [4, 5, 3]


# -----------------------------------------------
# STEP 3: Filter values by keep/remove lists
# -----------------------------------------------
# Flattens structured values and applies filtering logic.

def get_values_from_names(names_and_values: dict, val_keep: list, val_remove: list) -> list:
    names = names_and_values.get('names', [])
    vals = names_and_values.get('values', [])
    flat = []

    for i, v in enumerate(vals):
        name = names[i] if i < len(names) else f"name{i}"
        if isinstance(v, list):
            for sub in v:
                flat.append(f"{name}.{sub}")
        else:
            flat.append(f"{name}.{v}")

    if val_keep:
        return [f for f in flat if any(k in f for k in val_keep)]
    if val_remove:
        return [f for f in flat if all(r not in f for r in val_remove)]
    return flat

# # Test STEP 3
# sample = {
#     "names": ["country", "year"],
#     "values": [["USA", "UK"], [2020, 2021]]
# }
# print(get_values_from_names(sample, [], []))         # → ['country.USA', 'country.UK', 'year.2020', 'year.2021']
# print(get_values_from_names(sample, ['USA'], []))    # → ['country.USA']
# print(get_values_from_names(sample, [], ['UK']))     # → ['country.USA', 'year.2020', 'year.2021']


# -----------------------------------------------
# STEP 4: Sample consecutive values
# -----------------------------------------------
# Picks a block of values of random length within a given range.

def sample_values(val_list, val_sample):
    """
    Picks a block of values of random length within a given range.
    If val_sample is not a 2-element list/tuple, returns ALL values.
    """
    # “[0]” or any single-element val_sample → keep everything
    if not (isinstance(val_sample, (list, tuple)) and len(val_sample) == 2):
        return {'listAttrNames': val_list, 'isAggPossible': 'true'}

    total = len(val_list)
    n, m = val_sample
    # clamp n,m to [1,total]
    n = max(1, min(n, total))
    m = max(n, min(m, total))
    size = n if n == m else int(np.random.randint(n, m + 1))
    start = int(np.random.randint(0, total - size + 1))
    sel = val_list[start:start + size]
    is_agg = 'false' if len(sel) == 1 else 'true'
    return {'listAttrNames': sel, 'isAggPossible': is_agg}


# # Test STEP 4
# np.random.seed(123)
# val_list = ["A", "B", "C", "D", "E"]
# val_sample = [2, 3]
# print(sample_values(val_list, val_sample))
# # → sample with 2–3 consecutive elements from val_list


# -----------------------------------------------
# STEP 5: Sample hierarchical attributes
# -----------------------------------------------
# Applies step 4 logic recursively for each level in hierarchical names.

def sample_values_from_hierarchy(list_composite, val_sample):
    # Split each “A.B.C” into [A,B,C]
    split_rows = [row.split('.') for row in list_composite]
    depth = len(split_rows[0])

    # Build a matrix of rows
    matrix = [row[:] for row in split_rows]

    # Helper: preserve first‐seen order when uniquifying
    def unique_preserve_order(seq):
        seen = set()
        out = []
        for x in seq:
            if x not in seen:
                seen.add(x)
                out.append(x)
        return out

    # --------------------------
    # Level 0: sample among top‐level codes
    # --------------------------
    # Extract the 0th column in order
    tops = unique_preserve_order(row[0] for row in matrix)
    r0 = sample_values(tops, val_sample)
    agg_flags = [r0['isAggPossible']]

    # Keep only rows whose top‐level is in r0['listAttrNames']
    matrix = [row for row in matrix if row[0] in r0['listAttrNames']]

    # --------------------------
    # Levels 1..depth-1
    # --------------------------
    for lvl in range(1, depth):
        next_matrix = []
        lvl_flag = 'true'
        # for each distinct parent at level-(lvl-1)
        parents = unique_preserve_order(row[0] for row in matrix)
        for parent in parents:
            # collect all rows under that parent
            group = [row for row in matrix if row[0] == parent]
            # find the unique values at this level
            vals = unique_preserve_order(r[lvl] for r in group)
            r = sample_values(vals, val_sample)
            if r['isAggPossible'] == 'false':
                lvl_flag = 'false'
            # keep only the rows whose lvl value was picked
            allowed = set(r['listAttrNames'])
            next_matrix.extend([row for row in group if row[lvl] in allowed])
        agg_flags.append(lvl_flag)
        matrix = next_matrix

    # Reconstruct the “A.B.C” strings
    collapsed = ['.'.join(row) for row in matrix]
    return {
        'listAttrNames': collapsed,
        'isAggPossible': '.'.join(agg_flags)
    }


# # Test STEP 5
# composite_attrs = [
#     "Asia.China.Beijing", "Asia.China.Shanghai",
#     "Asia.India.Delhi", "Asia.India.Mumbai",
#     "Europe.France.Paris", "Europe.France.Lyon",
#     "Europe.Germany.Berlin", "Europe.Germany.Munich"
# ]
# val_sample = [1, 2]
# np.random.seed(123)
# print(sample_values_from_hierarchy(composite_attrs, val_sample))


# -----------------------------------------------
# STEP 6: Get index from attribute codes
# -----------------------------------------------
# Returns index of first attribute with matching code.

def get_attr_from_codes(attr_list, codes):
    if isinstance(codes, str):
        codes = [codes]
    for i, attr in enumerate(attr_list):
        if attr.get("code") in codes:
            return i
    return None

# # Test STEP 6
# attr_list_example = [
#     {"code": "x1", "names": ["Country"]},
#     {"code": "x2", "names": ["Year"]},
#     {"code": "x3", "names": ["City"]}
# ]
# print(get_attr_from_codes(attr_list_example, "x2"))  # → 1
# print(get_attr_from_codes(attr_list_example, "x5"))  # → None


# -----------------------------------------------
# STEP 7: Get codes by attribute name
# -----------------------------------------------
# Returns all codes where the name is included.

def get_codes_from_name(semantic_data, attr_name):
    codes = []
    for entry in semantic_data.get("data", []):
        if attr_name in entry.get("names", []):
            codes.append(entry["code"])
    return codes

# # Test STEP 7
# semantic_data = {
#     "data": [
#         {"code": "x1", "names": ["Country"]},
#         {"code": "x2", "names": ["Year"]},
#         {"code": "x3", "names": ["City"]}
#     ]
# }
# print(get_codes_from_name(semantic_data, "Year"))     # → ['x2']
# print(get_codes_from_name(semantic_data, "Unknown"))  # → []

